Sources/SwiftCoAP/SCMessage.swift

```diff
@@ -32,87 +34,171 @@ protocol SCCoAPTransportLayerDelegate: class {
+                guard let self = self else { return }
+                self.transportLayerDelegate?.transportLayerObject(self, didFailWithError: error as NSError)
+            case .setup:
+                os_log("Connection entered SETUP state", log: .default, type: .info)
+            case .waiting(let reason):
+                os_log("Connection entered WAITING state. Reason %@", log: .default, type: .info, reason.debugDescription)
+            case .preparing:
+                os_log("Connection entered PREPAIRING state", log: .default, type: .info)
+            case .ready:
+                os_log("Connection entered READY state", log: .default, type: .info)
+                guard let self = self else { return }
+                self.startReads(from: connection, withHostPort: hostPort)
+            case .cancelled:
+                os_log("Connection entered is CANCELLED", log: .default, type: .info)
+                self?.connections.removeValue(forKey: hostPort)
+            @unknown default:
+                os_log("Connection is in UNKNOWN state", log: .default, type: .info)
+            }
+        }
+        return connection
     }
-    
-    fileprivate func setUpUdpSocket() -> Bool {
-        udpSocket = GCDAsyncUdpSocket(delegate: self, delegateQueue: DispatchQueue.main)
-        do {
-            try udpSocket!.bind(toPort: port)
-            try udpSocket!.beginReceiving()
-        } catch {
-            return false
+
+    private func mustGetConnection(forHost host: String, port: UInt16) -> NWConnection {
+        os_log("Currently %d NWConnection object(s) are alive", connections.count)
+        os_log("Getting connection object for HOST %@, PORT %d", log: .default, type: .info, host, port)
+        let connectionKey = HostPortKey(host: host, port: port)
+        if let connection = connections[connectionKey], connection.state != .cancelled {
+            os_log("Reusing existing NWConnection for HOST %@, PORT %d", log: .default, type: .info, host, port)
+            return connection
+        }
+        let connection = NWConnection(host: NWEndpoint.Host(host), port: NWEndpoint.Port(rawValue: port)!, using: .udp)
+        connections[connectionKey] = connection
+        os_log("New NWConnection created for HOST %@, PORT %d", log: .default, type: .info, host, port)
+        return connection
+    }
+
+    private func startReads(from connection: NWConnection, withHostPort hostPort: HostPortKey) {
+        var waitingForNextDatagram = true
+        while connection.state == .ready {
+            if waitingForNextDatagram {
+                waitingForNextDatagram = false
+                connection.receiveMessage { [weak self] data, context, complete, error in
+                    waitingForNextDatagram = true
+                    guard let self = self else { return }
+                    if error != nil {
+                        self.transportLayerDelegate.transportLayerObject(self, didFailWithError: error! as NSError)
```

Comment:
```suggestion
                        self.transportLayerDelegate.transportLayerObject(self, didFailWithError: error)
```
Let's be a little safer.